sentry-android-core/src/main/java/io/sentry/android/core/DefaultAndroidEventProcessor.java

```diff
@@ -1,20 +1,742 @@
+
+    BufferedReader br = null;
+    try {
+      File file = new File("/proc/version");
+      if (!file.canRead()) {
+        return defaultVersion;
+      }
+
+      br = new BufferedReader(new FileReader(file));
+      return br.readLine();
+    } catch (Exception e) {
+      log(SentryLevel.ERROR, errorMsg, e);
+    } finally {
+      if (br != null) {
+        try {
+          br.close();
+        } catch (IOException ioe) {
+          log(SentryLevel.ERROR, errorMsg, ioe);
+        }
+      }
+    }
+
+    return defaultVersion;
+  }
+
+  /**
+   * Attempt to discover if this device is currently rooted. From:
+   * https://stackoverflow.com/questions/1101380/determine-if-running-on-a-rooted-device
+   *
+   * @return true if heuristics show the device is probably rooted, otherwise false
+   */
+  private Boolean isRooted() {
+    // we could get some inspiration from https://github.com/scottyab/rootbeer
+    if (Build.TAGS != null && Build.TAGS.contains("test-keys")) {
+      return true;
+    }
+
+    String[] probableRootPaths = {
+      "/data/local/bin/su",
+      "/data/local/su",
+      "/data/local/xbin/su",
+      "/sbin/su",
+      "/su/bin",
+      "/su/bin/su",
+      "/system/app/SuperSU",
+      "/system/app/SuperSU.apk",
+      "/system/app/Superuser",
+      "/system/app/Superuser.apk",
+      "/system/bin/failsafe/su",
+      "/system/bin/su",
+      "/system/sd/xbin/su",
+      "/system/xbin/daemonsu",
+      "/system/xbin/su"
+    };
+
+    for (String probableRootPath : probableRootPaths) {
+      try {
+        if (new File(probableRootPath).exists()) {
+          return true;
+        }
+      } catch (Exception e) {
+        log(
+            SentryLevel.ERROR,
+            "Exception while attempting to detect whether the device is rooted",
+            e);
+      }
+    }
+
+    return false;
+  }
+
+  /**
+   * Get the human-facing Application name.
+   *
+   * @return Application name
+   */
+  private String getApplicationName() {
+    // possible to get also from
+    // context.getPackageManager().getApplicationLabel(context.getPackageManager().getApplicationInfo(context.getPackageName(), 0))
+    // should we fallback?
+    try {
+      ApplicationInfo applicationInfo = context.getApplicationInfo();
+      int stringId = applicationInfo.labelRes;
+      if (stringId == 0) {
+        if (applicationInfo.nonLocalizedLabel != null) {
+          return applicationInfo.nonLocalizedLabel.toString();
+        }
+      } else {
+        return context.getString(stringId);
+      }
+    } catch (Exception e) {
+      log(SentryLevel.ERROR, "Error getting application name.", e);
+    }
+
+    return null;
+  }
+
+  public User getUser() {
+    User user = new User();
+
+    user.setId(getAndroidId());
+
+    return user;
+  }
+
+  @SuppressWarnings("HardwareIds")
+  private String getAndroidId() {
+    // Android 29 has changed and -> Avoid using hardware identifiers, find another way in the
```

Comment:
What did Android 29 change? Don't follow why we need to find another way

Reply:
for now, it's too early to deal with this problem, but important to point out.
https://developer.android.com/training/articles/user-data-ids

one important thing is to test permission `READ_PRIVILEGED_PHONE_STATE`